sphericart/include/macros.hpp

```diff
@@ -35,7 +35,17 @@
 // double precision, as that slows down greatly floating-point execution.
 // basically this static casts the constant to the type of the target
 // pointer
-#define cast(sph, lit) static_cast<std::remove_pointer_t<decltype(sph)>>(lit)
+template <typename T> struct remove_pointer {
+    using type = T;
+};
+
+template <typename T> struct remove_pointer<T*> {
+    using type = T;
+};
+
+template <typename T> using remove_pointer_t = typename remove_pointer<T>::type;
```

Comment:
Could this use the version in `<type_traits>`? https://en.cppreference.com/w/cpp/types/remove_pointer

Reply:
Unfortunately not. If I use what was originally there, I get this:

```>       return self.calculator.compute(xyz)
E       RuntimeError: KernelFactory::compileAndCacheKernel: Failed to compile CUDA program:
E       sphericart_impl.cu(1480): error: namespace "std" has no member "remove_pointer_t"
E                     HARDCODED_SPH_MACRO(1, x, y, z, x2, y2, z2, buffer_sph, get_index);
E                     ^
E       
E       sphericart_impl.cu(1480): error: expected a ">"
E                     HARDCODED_SPH_MACRO(1, x, y, z, x2, y2, z2, buffer_sph, get_index);
E                     ^
E       
E       sphericart_impl.cu(1480): error: expected a "("
E                     HARDCODED_SPH_MACRO(1, x, y, z, x2, y2, z2, buffer_sph, get_index);
E                     ^
E       
E       sphericart_impl.cu(1480): error: expected an expression
E                     HARDCODED_SPH_MACRO(1, x, y, z, x2, y2, z2, buffer_sph, get_index);
E                     ^
E       
E       sphericart_impl.cu(1480): error: expected a ")"
E                     HARDCODED_SPH_MACRO(1, x, y, z, x2, y2, z2, buffer_sph, get_index);
E                     ^
E       
E       sphericart_impl.cu(1482): error: namespace "std" has no member "remove_pointer_t"
E                         HARDCODED_SPH_DERIVATIVE_MACRO(
E                         ^
E       
E       sphericart_impl.cu(1482): error: expected a ">"
E                         HARDCODED_SPH_DERIVATIVE_MACRO(
E                         ^
E       
E       sphericart_impl.cu(1482): error: expected a "("
E                         HARDCODED_SPH_DERIVATIVE_MACRO(
E                         ^
E       
E       sphericart_impl.cu(1482): error: expected an expression
E                         HARDCODED_SPH_DERIVATIVE_MACRO(
E                         ^
E       
E       sphericart_impl.cu(1482): error: expected a ")"
E                         HARDCODED_SPH_DERIVATIVE_MACRO(
E                         ^
E       
E       sphericart_impl.cu(1488): error: namespace "std" has no member "remove_pointer_t"
E                         HARDCODED_SPH_SECOND_DERIVATIVE_MACRO(
E                         ^
E       
E       sphericart_impl.cu(1488): error: expected a ">"
E                         HARDCODED_SPH_SECOND_DERIVATIVE_MACRO(
E                         ^
E       
E       sphericart_impl.cu(1488): error: expected a "("
E                         HARDCODED_SPH_SECOND_DERIVATIVE_MACRO(
E                         ^
E       
E       sphericart_impl.cu(1488): error: expected an expression
E                         HARDCODED_SPH_SECOND_DERIVATIVE_MACRO(
E                         ^
E       
E       sphericart_impl.cu(1488): error: expected a ")"
E                         HARDCODED_SPH_SECOND_DERIVATIVE_MACRO(
E                         ^
E       
E       sphericart_impl.cu(1504): error: namespace "std" has no member "remove_pointer_t"
E                     COMPUTE_SPH_L0(buffer_sph, get_index);
E                     ^
E       
E       sphericart_impl.cu(1504): error: expected a ">"
E                     COMPUTE_SPH_L0(buffer_sph, get_index);
E                     ^
E       
E       sphericart_impl.cu(1504): error: expected a "("
E                     COMPUTE_SPH_L0(buffer_sph, get_index);
E                     ^
E       
E       sphericart_impl.cu(1504): error: expected an expression
E                     COMPUTE_SPH_L0(buffer_sph, get_index);
E                     ^
E       
E       sphericart_impl.cu(1504): error: expected a ")"
E                     COMPUTE_SPH_L0(buffer_sph, get_index);
E                     ^
E       
E       sphericart_impl.cu(1506): error: namespace "std" has no member "remove_pointer_t"
E                         COMPUTE_SPH_DERIVATIVE_L0(
E                         ^
E       
E       sphericart_impl.cu(1506): error: expected a ">"
E                         COMPUTE_SPH_DERIVATIVE_L0(
E                         ^
E       
E       sphericart_impl.cu(1506): error: expected a "("
E                         COMPUTE_SPH_DERIVATIVE_L0(
E                         ^
E       
E       sphericart_impl.cu(1506): error: expected an expression
E                         COMPUTE_SPH_DERIVATIVE_L0(
E                         ^
E       
E       sphericart_impl.cu(1506): error: expected a ")"
E                         COMPUTE_SPH_DERIVATIVE_L0(
E                         ^
E       
E       sphericart_impl.cu(1511): error: namespace "std" has no member "remove_pointer_t"
E                             COMPUTE_SPH_SECOND_DERIVATIVE_L0(
E                             ^
E       
E       sphericart_impl.cu(1511): error: expected a ">"
E                             COMPUTE_SPH_SECOND_DERIVATIVE_L0(
E                             ^
E       
E       sphericart_impl.cu(1511): error: expected a "("
E                             COMPUTE_SPH_SECOND_DERIVATIVE_L0(
E                             ^
E       
E       sphericart_impl.cu(1511): error: expected an expression
E                             COMPUTE_SPH_SECOND_DERIVATIVE_L0(
E                             ^
E       
E       sphericart_impl.cu(1511): error: expected a ")"
E                             COMPUTE_SPH_SECOND_DERIVATIVE_L0(
E                             ^
E       
E       30 errors detected in the compilation of "sphericart_impl.cu".

../../miniconda3/envs/CSCS/lib/python3.10/site-packages/sphericart/torch/__init__.py:176: RuntimeError```